ngraph/core/src/op/einsum.cpp

```diff
@@ -0,0 +1,299 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <algorithm>
+#include <cctype>
+#include <ngraph/validation_util.hpp>
+#include <string>
+#include <unordered_map>
+
+#include "itt.hpp"
+#include "ngraph/op/einsum.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v7::Einsum, "Einsum", 7);
+
+op::v7::Einsum::Einsum(const OutputVector& inputs, const std::string& equation)
+    : Op(inputs)
+    , m_equation(equation)
+{
+    // normalize input equation by removing extra white-spaces from the equation
+    m_equation.erase(std::remove_if(m_equation.begin(), m_equation.end(), ::isspace),
+                     m_equation.end());
+
+    constructor_validate_and_infer_types();
+}
+
+bool op::v7::Einsum::is_subscript_correct(const std::string& subscript, bool& is_ellipsis_met)
+{
+    is_ellipsis_met = false;
+    auto subscript_length = subscript.length();
+    for (size_t ch_idx = 0; ch_idx < subscript_length; ++ch_idx)
+    {
+        if (is_ellipsis_met == false && ((subscript_length - ch_idx) > 2) &&
+            (subscript.substr(ch_idx, 3).compare("...") == 0))
+        {
+            // mark that ellipsis is met once
+            is_ellipsis_met = true;
+
+            // make additional increment since ellipsis consists of three dots.
+            ch_idx += 2;
+        }
+        else if (std::isalpha(subscript[ch_idx]) == 0)
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void op::v7::Einsum::parse_equation(const std::string& equation,
+                                    std::vector<std::string>& input_subscripts,
+                                    std::string& output_subscript)
+{
+    NGRAPH_OP_SCOPE(v7_Einsum_parse_equation);
+
+    // split equation to input subscripts and an output subscript
+    auto pos_output_delimeter = equation.find("->");
+    auto input_subscripts_str = equation.substr(0, pos_output_delimeter);
+
+    // split the input subscripts into a vector of input subscripts
+    bool is_ellipsis_met = false;
+    input_subscripts.clear();
+    std::istringstream input;
+    input.str(input_subscripts_str);
+    for (std::string input_subscript; std::getline(input, input_subscript, ',');)
+    {
+        bool local_is_ellipsis_met = false;
+        // check that input subscript contains only alphabetic letter or ellipsis
+        NGRAPH_CHECK(is_subscript_correct(input_subscript, local_is_ellipsis_met) == true,
```

Comment:
You can simplify
```suggestion
        NGRAPH_CHECK(is_subscript_correct(input_subscript, local_is_ellipsis_met),
```

Reply:
I also don't like `== true` statements. The same comment for line 78

Reply:
I prefer an explicit form to expect a concrete returned type even for boolean function

Reply:
I prefer explicit form to expect a concrete type in return even for boolean function. It can prevent some incorrect changes in this function signature.